ospf6d/ospf6_asbr.c

```diff
@@ -2153,25 +2156,65 @@ static const struct route_map_rule_cmd
 	ospf6_routemap_rule_set_metric_type_free,
 };
 
+struct ospf6_metric {
+	enum { metric_increment, metric_decrement, metric_absolute } type;
+	bool used;
+	uint32_t metric;
+};
+
 static enum route_map_cmd_result_t
 ospf6_routemap_rule_set_metric(void *rule, const struct prefix *prefix,
 			       void *object)
 {
-	char *metric = rule;
-	struct ospf6_route *route = object;
+	struct ospf6_metric *metric;
+	struct ospf6_route *route;
+
+	/* Fetch routemap's rule information. */
+	metric = rule;
+	route = object;
+
+	/* Set metric out value. */
+	if (!metric->used)
+		return RMAP_OKAY;
+
+	route->path.cost = route->path.redistribute_cost;
+
+	if (metric->type == metric_increment)
+		route->path.cost += metric->metric;
+	else if (metric->type == metric_decrement)
+		route->path.cost -= metric->metric;
+	else if (metric->type == metric_absolute)
+		route->path.cost = metric->metric;
 
-	route->path.cost = atoi(metric);
 	return RMAP_OKAY;
 }
 
 static void *ospf6_routemap_rule_set_metric_compile(const char *arg)
 {
-	uint32_t metric;
-	char *endp;
-	metric = strtoul(arg, &endp, 0);
-	if (metric > OSPF_LS_INFINITY || *endp != '\0')
-		return NULL;
-	return XSTRDUP(MTYPE_ROUTE_MAP_COMPILED, arg);
+	struct ospf6_metric *metric;
+
+	metric = XCALLOC(MTYPE_ROUTE_MAP_COMPILED, sizeof(*metric));
+	metric->used = false;
+
+	if (all_digit(arg))
+		metric->type = metric_absolute;
+
+	if ((arg[0] == '+') && all_digit(arg + 1)) {
+		metric->type = metric_increment;
+		arg++;
+	}
+
+	if ((arg[0] == '-') && all_digit(arg + 1)) {
+		metric->type = metric_decrement;
+		arg++;
+	}
+
+	metric->metric = strtoul(arg, NULL, 10);
```

Comment:
Does this work with a metric subtraction since this could be misinterpreted?  Shouldn't you skip the +/- so we don't have to wonder about the precise strtoul() behavior where "u" stands for unsigned.

Reply:
Skipping the +/- is exactly what this code is doing. strtoul() will only see a number with no sign beforehand.
btw this is the same code as ospfd is using.